controller_manager/src/controller_manager.cpp

```diff
@@ -53,61 +64,772 @@ ControllerManager::load_controller(
+    auto stop_list_it = std::find(
+      stop_request_.begin(), stop_request_.end(), controller.c);
+    bool in_stop_list = stop_list_it != stop_request_.end();
+
+    auto start_list_it = std::find(
+      start_request_.begin(), start_request_.end(), controller.c);
+    bool in_start_list = start_list_it != start_request_.end();
+
+    const bool is_running = is_controller_running(*controller.c);
+
+    auto handle_conflict = [&](const std::string & msg)
+      {
+        if (strictness == controller_manager_msgs::srv::SwitchController::Request::STRICT) {
+          RCLCPP_ERROR_STREAM(
+            get_logger(),
+            msg);
+          stop_request_.clear();
+          start_request_.clear();
+          return controller_interface::return_type::ERROR;
+        }
+        RCLCPP_DEBUG_STREAM(
+          get_logger(),
+          "Could not stop controller '" << controller.info.name <<
+            "' since it is not running");
+        return controller_interface::return_type::SUCCESS;
+      };
+    if (!is_running && in_stop_list) {      // check for double stop
+      auto ret = handle_conflict(
+        "Could not stop controller '" + controller.info.name +
+        "' since it is not running");
+      if (ret != controller_interface::return_type::SUCCESS) {
+        return ret;
+      }
+      in_stop_list = false;
+      stop_request_.erase(stop_list_it);
+    }
+
+    if (is_running && !in_stop_list && in_start_list) {  // check for doubled start
+      auto ret = handle_conflict(
+        "Could not start controller '" + controller.info.name +
+        "' since it is already running");
+      if (ret != controller_interface::return_type::SUCCESS) {
+        return ret;
+      }
+      in_start_list = false;
+      start_request_.erase(start_list_it);
+    }
+
+#ifdef TODO_IMPLEMENT_RESOURCE_CHECKING
+    if (is_running && in_stop_list && !in_start_list) {  // running and real stop
+      switch_stop_list_.push_back(info);
+    } else if (!is_running && !in_stop_list && in_start_list) {  // start, but no restart
+      switch_start_list_.push_back(info);
+    }
+
+    bool add_to_list = is_running;
+    if (in_stop_list) {
+      add_to_list = false;
+    }
+    if (in_start_list) {
+      add_to_list = true;
+    }
+
+    if (add_to_list) {
+      info_list.push_back(info);
+    }
+#endif
+  }
+
+#ifdef TODO_IMPLEMENT_RESOURCE_CHECKING
+  bool in_conflict = robot_hw_->checkForConflict(info_list);
+  if (in_conflict) {
+    RCLCPP_ERROR(get_logger(), "Could not switch controllers, due to resource conflict");
+    stop_request_.clear();
+    start_request_.clear();
+    return controller_interface::return_type::ERROR;
+  }
+
+  if (!robot_hw_->prepareSwitch(switch_start_list_, switch_stop_list_)) {
+    RCLCPP_ERROR(
+      get_logger(),
+      "Could not switch controllers. The hardware interface combination "
+      "for the requested controllers is unfeasible.");
+    stop_request_.clear();
```

Comment:
why does this have to be a class member variable?

Reply:
It's populated from a service thread and used from the RT thread in `switch_controller()` and used by the RT thread in `stop_controllers()`.

It could be moved inside the `SwitchParams` struct, but it would still be part of a member variable.